service/src/main/kotlin/fi/espoo/evaka/dailyservicetimes/DailyServiceTimesController.kt

```diff
@@ -109,61 +171,98 @@ class DailyServiceTimesController(
-                            DateRange(range.end.plusDays(1), it.validityPeriod.end)
-                        else
-                            throw IllegalStateException("Unsupported overlap")
-                    )
+        val overlapping = tx.getOverlappingChildDailyServiceTimes(childId, new)
+        overlapping.forEach { (oldId, old) ->
+            if (new.contains(old)) {
+                tx.deleteChildDailyServiceTimes(oldId)
+            } else {
+                val updatedRange = when {
+                    // old is finite and old contains new in the middle
+                    //
+                    // old OOOOOO
+                    // new   NNN
+                    // --> error
+                    old.contains(new) && old.end != null && old.start != new.start && old.end != new.end ->
+                        throw Conflict("Unsupported overlap")
+
+                    // old contains new in the start
+                    //
+                    // old OOOOOO
+                    // new NNN
+                    // --> NNNOOO
+                    new.end != null && old.start == new.start ->
+                        DateRange(new.end.plusDays(1), old.end)
+
+                    // old overlaps with the end of new
+                    //
+                    // old   OOOO
+                    // new NNN
+                    // --> NNNOOO
+                    new.end != null && old.end != null && old.includes(new.end) ->
+                        DateRange(new.end.plusDays(1), old.end)
+
+                    // old contains new in the end
+                    //
+                    // old OOOOOO
+                    // new    NNN
+                    // --> OOONNN
+                    old.end == null || old.end == new.end ->
+                        DateRange(old.start, new.start.minusDays(1))
+
+                    // old overlaps with the beginning of new
+                    //
+                    // old OOOO
+                    // new    NNN
+                    // --> OOONNN
+                    old.includes(new.start) ->
+                        DateRange(old.start, new.start.minusDays(1))
+
+                    else ->
+                        throw Conflict("Unsupported overlap")
                 }
+                tx.updateChildDailyServiceTimesValidity(oldId, updatedRange)
             }
         }
     }
 
     private fun deleteCollidingReservationsAndNotify(
         tx: Database.Transaction,
+        now: HelsinkiDateTime,
         id: DailyServiceTimesId,
-        evakaClock: EvakaClock
+        childId: ChildId,
+        validityPeriod: DateRange,
     ) {
-        val dst = tx.getChildDailyServiceTimeValidity(id) ?: return
-
-        if (dst.validityPeriod.end?.isBefore(evakaClock.today()) == true) {
-            return
-        }
+        val today = now.toLocalDate()
+        if ((validityPeriod.end ?: LocalDate.MAX) <= today) return
```

Comment:
Isn't this basically impossible at the moment? I think throwing an error would be preferable.

Reply:
Fixed 👍